dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/validation/validator/relationship/MandatoryFieldsValidatorTest.java

```diff
@@ -126,23 +129,109 @@ void verifyRelationshipValidationFailsOnMissingFrom()
 
         validator.validate( reporter, bundle, relationship );
 
-        assertMissingProperty( reporter, relationship, "from" );
+        assertHasError( reporter, relationship, E4001,
+            "Relationship item `from` for relationship `" + relationshipUid
```

Comment:
I would only assert on `"Relationship item from for relationship " + relationshipUid` so that you don't have to change every test if you want to tweak the message.

I think

`is invalid: an item must link exactly one of trackedEntity, enrollment, event.`

is missing a "to" before "link" in

`is invalid: an item must link to exactly one of trackedEntity, enrollment, event.`